File: concordion/src/test/java/nz/co/twoten/driver/google/GoogleResultsPage.java
/**
 * Code from https://bitbucket.org/twoten/opensourcetesting
 */
package nz.co.twoten.driver.google;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

/**
 * A WebDriver Page Object corresponding to the Google Results Page.
 */
public class GoogleResultsPage {
	
	@FindBy(id = "res") 
	private WebElement resultWrapper;

	@FindBy(className = "l")
	private WebElement firstResultLink;
	
    @FindBy(className = "r")
    private WebElement resultLink;

    private final WebDriver driver;

	/**
	 * Initialises the results page and waits for the page to fully load.
	 * Assumes that the results page is already loading.
	 */
    public GoogleResultsPage(WebDriver driver) {
		this.driver = driver;
        PageFactory.initElements(driver, this);
        waitForFooter();
	}

	/**
	 * Checks whether the specified text occurs in any result on the results page.
	 */
    public boolean resultsContain(String text) {
		List<WebElement> resultsText = resultWrapper.findElements(By.className("s"));
		for (WebElement result : resultsText) {
            if (result.getText().contains(text)) {
                return true;
            }
        }
		return false;
	}

	/**
	 * Returns the text of the topmost result from the results page.
	 */
    public String getTopResultTitle() {
		return firstResultLink.getText();
	}
	
    /**
     * Returns the text of the calculator result from the results page.
     */
    public String getResult() {
        return resultLink.getText();
    }
    
    private void waitForFooter() {
		WebDriverWait wait = new  WebDriverWait(driver, 30);
		wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("foot")));
	}
}

File: concordion.net/src/concordion.spec/Provider/GoogleResultsPage.cs
/**
 * Code ported from java from https://bitbucket.org/twoten/opensourcetesting
 * todd@goneopen.com 2012
 */

using System;
using System.Linq;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace Concordion.Spec.Provider
{
    public class GoogleResultsPage
    {
        private readonly By _results = By.Id("res");
        private readonly By _headings = By.Id("s");
        private readonly By _footer = By.Id("foot");
        private readonly By _links = By.ClassName("r");

        private readonly IWebDriver _webDriver;
 
        public GoogleResultsPage(IWebDriver webDriver)
        {
            _webDriver = webDriver;
            waitForFooter();
        }

        public bool ResultsContain(string text)
        {
            var resultsText = _webDriver.FindElement(_results).FindElements(_headings);
            return resultsText.Any(result => result.Text.Contains(text));
        }

        public string Results()
        {
            return _webDriver.FindElement(_links).Text;
        }

        private void waitForFooter()
        {
            var wait = new WebDriverWait(_webDriver, TimeSpan.FromSeconds(30));
            wait.Until(ExpectedConditions.ElementIsVisible(_footer));
        }
    }
}
